OMPL Constrained State Space Planning

JeroenDM opened the pull request:
### Description

In the context of a [Google Summer of Code project](https://github.com/ros-planning/moveit/issues/2092), I would like to add the constrained planning capabilities of OMPL to the existing interface. Previously, I split up the code in several pull requests. This was not entirely clear and impractical for testing. Hence this new approach, a single pull request that can be reviewed on a per commit basis.

**Edit** this pull request is quite large, but can be made smaller by first merging these tests separately:
- Add test to ompl interface for StateValidityChecker #2247 (already approved)
- WIP: add test for PlanningContextManager in ompl interface #2248 (needs some discussion)

### Why
It is one of the projects on the [future projects page](https://moveit.ros.org/documentation/contributing/future_projects/).

It can solve the long-standing [joint flip bug](https://github.com/ros-planning/moveit/issues/562) when planning with path constraints. It is especially useful in cases where the constrained region has zero volume in Cartesian space and the ` enforce_joint_model_state_space` (rejection sampling) does not work. For example planning with the end-effector constrained to a plane:

![case_2](https://user-images.githubusercontent.com/11537861/90881322-3d437d00-e3aa-11ea-8660-1084b8c5c93c.gif)

Or a line:

![case_3](https://user-images.githubusercontent.com/11537861/90881337-43d1f480-e3aa-11ea-94ee-297da2583a05.gif)

The current default planner that uses the `PoseModelStateSpace` can solve these problems, but the solution will often contain large jumps in joint space because it plans in Cartesian space. (Especially when adding obstacles in the scene.) This new planner plans in joint space. The example below illustrates the difference.

With the default `PoseModelStateSpace`:

![case_4_pose_model](https://user-images.githubusercontent.com/11537861/90881354-4df3f300-e3aa-11ea-930b-a732757f3957.gif)

With OMPL's `ConstrainedStateSpace`:

![case_4](https://user-images.githubusercontent.com/11537861/90881359-52b8a700-e3aa-11ea-9563-68fc4cb2e81c.gif)

Finally (and maybe most importantly) once this new planning approach is integrated, the possibilities for adding new constraints are endless. It would be really nice if we could pass arbitrary user-defined constraints to OMPL through some C++ interface in the future.

### Checklist

- [x] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [x] [Extend the tutorials.](https://github.com/ros-planning/moveit_tutorials/pull/518)
- [ ] Document API changes relevant to the user in the [MIGRATION.md](https://github.com/ros-planning/moveit/blob/master/MIGRATION.md) notes
- [x] Create tests, which fail without this PR [reference](https://ros-planning.github.io/moveit_tutorials/doc/tests/tests_tutorial.html)
- [ ] Include a screenshot if changing a GUI
- [ ] While waiting for someone to review your request, please help review [another open pull request](https://github.com/ros-planning/moveit/pulls) to support the maintainers

JeroenDM commented:
The `kinetic` build fails with:
```
/root/ros_ws/src/moveit/moveit_planners/ompl/ompl_interface/include/moveit/ompl_interface/model_based_planning_context.h:48:63: fatal error: ompl/base/spaces/constraint/ConstrainedStateSpace.h: No such file or directory
```
I suspect this is because kinetic is using a version of OMPL from before [this commit](https://github.com/ompl/ompl/commit/958eceb3618b3b679872c8499f6baac0eb5f5418).

felixvd commented on moveit_planners/ompl/ompl_interface/include/moveit/ompl_interface/parameterization/joint_space/constrained_planning_state_space.h:
```diff
@@ -0,0 +1,81 @@
+ * ConstrainedPlanningStateSpace.
+ *
+ * This class overrides the virtual methods to copy states from OMPL to MoveIt and vice-versa. It is used as state space
+ * in the `ModelBasedPlanningContextSpecification` and can handle general `ompl::base::State*` pointers that need to be
```
On first read I am not sure what is part of MoveIt and what is part of OMPL.

JeroenDM replied:
I feel like I'm too deep in all of this to understand the question :s

In the comments and code, everything from OMPL is prefixed with `ompl::base`. In the arguments of the methods of this class, OMPL states are called `ompl_state` and MoveIt states fo type `RobotState` are called `robot_state`.

felixvd commented on moveit_planners/ompl/ompl_interface/include/moveit/ompl_interface/parameterization/joint_space/constrained_planning_state_space_factory.h:
```diff
@@ -0,0 +1,64 @@
+
+  /** \brief Return a priority that this planner should be used for this specific planning problem.
+   *
+   * This state space factory is currently only used if `use_ompl_constrained_state_space` was set to `true` in
```
Might be worth linking the discussion about this here

JeroenDM replied:
Do you mean [this writeup](https://github.com/JeroenDM/moveit/pull/2)?

felixvd commented on moveit_planners/ompl/ompl_interface/src/detail/ompl_constraints.cpp:
```diff
@@ -0,0 +1,318 @@
+    }
+  }
+
+  ROS_INFO_STREAM_NAMED(LOGNAME, "X dimension constraint? " << is_dim_constrained_[0]);
```
This should probably be prettier and DEBUG level

JeroenDM replied:
done

felixvd commented on moveit_planners/ompl/ompl_interface/src/detail/ompl_constraints.cpp:
```diff
@@ -0,0 +1,318 @@
+  std::size_t num_ori_con = constraints.orientation_constraints.size();
+
+  // Only positions constraints are actually created below.
+  // All the other code is just given feedback to the user.
```
?

JeroenDM replied:
The whole factory method is a template code to support different constraints, but only a single option is actually returns a non-nullptr at the moment (position constraints). All the other cases spit out warnings or errors.

I guess I can remove all the unnecessary template code and only add it when needed, I never though about that.

JeroenDM replied:
I updated the comment and left the rest of the function as is. I feel it is pretty clear with the log messages and allows for easy integration of orientation constraints and a combination of position and orientation constraints in the future.

Now the comment reads:
```C++
// The factory method contains template code to support different constraints, but only position constraints are
// currently supported. The other options return a nullptr for now and should not be used.
```

felixvd commented on moveit_planners/ompl/ompl_interface/src/detail/state_validity_checker.cpp:
```diff
@@ -194,4 +196,112 @@ double ompl_interface::StateValidityChecker::clearance(const ompl::base::State*
+    return state->as<ModelBasedStateSpace::StateType>()->isMarkedValid();
+
+  // do not use the unwrapped state here, as satisfiesBounds expects a state of type ConstrainedStateSpace::StateType
+  if (!si_->satisfiesBounds(wrapped_state))
```
Can we name `si_` to something more descriptive?

JeroenDM replied:
It's from the OMPL source code (`ompl::base::StateValidityChecker`). I could create a pull request there, but I can imagine it would be quite a large one.

I can also add a comment above every use here, it's only used 4 times.

felixvd commented on moveit_planners/ompl/ompl_interface/src/parameterization/joint_space/constrained_planning_state_space.cpp:
```diff
@@ -0,0 +1,108 @@
+  return ompl_state->as<ompl_interface::ConstrainedPlanningStateSpace::StateType>()->values + index;
+
+  // (jeroendm) left in the debug code, because it is not unlikely that issues will arise here when adding new things in
+  // the future. Use dynamic casting to make our debugging live easier.
```
You can also link to a writeup with debugging instructions instead

JeroenDM replied:
Good idea, it is a bit ugly to leave this in.

JeroenDM replied:
I replaced the long comment with: `// Developer tip: replace this with a dynamic_cast for debugging`. As for debugging instructions, I think it will be useful if I write a good description of all these details in a central place somewhere. But I need to come up with a good format and a good place to put it. Maybe the MoveIt [concepts page](https://moveit.ros.org/documentation/concepts/) on the website?

felixvd reviewed (commented):
I haven't even gotten halfway through skimming this. Just dropping some typo/style comments.

JeroenDM commented on moveit_planners/ompl/ompl_interface/include/moveit/ompl_interface/detail/ompl_constraints.h:
```diff
@@ -0,0 +1,326 @@
+
+/** \brief Represents upper and lower bound on a scalar value (double).
+ *
+ * This is necessary because an OMPL ConstrainedStateSpace requires a model of the constriants given as a generic
```
I started using that plugin this week :) I really needed it...

JeroenDM commented on moveit_planners/ompl/ompl_interface/include/moveit/ompl_interface/detail/ompl_constraints.h:
```diff
@@ -0,0 +1,326 @@
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of Willow Garage nor the names of its
```
Done

JeroenDM commented on moveit_planners/ompl/ompl_interface/include/moveit/ompl_interface/detail/ompl_constraints.h:
```diff
@@ -0,0 +1,326 @@
+/******************************************
+ * Position constraints
+ * ****************************************/
+/** \brief Box shaped position constraints
```
Good idea, that's a better name indeed.

JeroenDM replied:
done

JeroenDM commented:
@felixvd Thank you for the review! I addressed most of the small comments and fixes. I left some comments unresolved as I added a change that does not necessarily reflect you're proposed change but still resolves it in a reasonable way I think.

felixvd commented:
I hope I can get back to it soon, I didn't get even get through half!

JeroenDM commented:
Yeah, it's a lot. I still think my original approach to split the pull request into smaller parts was easier to review. But it's difficult to keep everything organized that way.

Anyway, I would ignore the test code for now and focus on the functional code. (Although the tests can serve a bit as documentation.)

lorepieri8 commented:
Is this problem present also in Moveit2?

JeroenDM commented:
@lorepieri8 Yes. As far as I know, MoveIt2 uses exactly the same OMPL interface code as MoveIt1.

mamoll reviewed (changes requested):
Looking good, Jeroen! I have made several small suggestions. The only one that might be a bit more work is the one related to refactoring the Bounds class.

JeroenDM commented on moveit_planners/ompl/ompl_interface/include/moveit/ompl_interface/detail/ompl_constraints.h:
```diff
@@ -0,0 +1,325 @@
+ * ****************************************/
+/** \brief Equality constraints on a link's position.
+ *
+ *  When you set the name of a constraint to 'use_equality_constraints', all constraints with a dimension lower that
```
I used the "name" field in [moveit_msgs/Constraints](http://docs.ros.org/latest/api/moveit_msgs/html/msg/Constraints.html), but it is hacky.

**I need to remove this comment** as it will not be necessary anymore if I add change the orientation constraint message as discussed in [this issue](https://github.com/ros-planning/moveit_msgs/issues/89). I started working on it but my pace is not the same as it was this summer...

JeroenDM commented on moveit_planners/ompl/ompl_interface/include/moveit/ompl_interface/detail/state_validity_checker.h:
```diff
@@ -81,4 +81,60 @@ class StateValidityChecker : public ompl::base::StateValidityChecker
+class ConstrainedPlanningStateValidityChecker : public StateValidityChecker
+{
+public:
+  using StateValidityChecker::isValid;
```
Because the `ompl::base::State` pointer passed to the state validity checker has a different type when planning in constrained state space. I need to "unwrap" it to get to the actual state, as is done [here](https://github.com/JeroenDM/moveit/blob/ad0aa16fbb206481fb1394c804176f509b84e2da/moveit_planners/ompl/ompl_interface/src/detail/state_validity_checker.cpp#L209) like so:
```C++
  // Unwrap the state from a ConstrainedStateSpace::StateType
  auto state = wrapped_state->as<ompl::base::ConstrainedStateSpace::StateType>()->getState();
```

I considered using dynamic casting to check at runtime what kind of pointer it is, but I assumed this would be slower and collision checking is often *the* bottleneck. (I did try this out and it works, but I did not benchmark it...).

Both solutions are equally ugly in my opinion...

JeroenDM commented:
@mamoll thank you for the detailed review! I'll do my best to find some time in the coming weeks to move this forward.

JeroenDM commented:
I created a [public Trello board](https://trello.com/b/TVCDP0Mx/moveit) to follow progress on this. Hopefully, this planning helps me to spend the little time I have to work on this effectively.

tylerjw commented on moveit_planners/ompl/ompl_interface/src/detail/state_validity_checker.cpp:
```diff
@@ -194,4 +196,112 @@ double ompl_interface::StateValidityChecker::clearance(const ompl::base::State*
+bool ompl_interface::ConstrainedPlanningStateValidityChecker::isValid(const ompl::base::State* wrapped_state,
+                                                                      double& dist, bool verbose) const
+{
+  // Unwrap the state from a ConstrainedStateSpace::StateType
```
assert that wrapped_stae is not nullptr

JeroenDM replied:
I also added it to the existing state validity checker.

tylerjw commented on moveit_planners/ompl/ompl_interface/src/detail/state_validity_checker.cpp:
```diff
@@ -194,4 +196,112 @@ double ompl_interface::StateValidityChecker::clearance(const ompl::base::State*
+      verbose ? collision_request_with_distance_verbose_ : collision_request_with_distance_, res, *robot_state);
+  dist = res.distance;
+  return !res.collision;
 }
```
newline at eof

tylerjw commented on moveit_planners/ompl/ompl_interface/src/model_based_planning_context.cpp:
```diff
@@ -116,11 +117,23 @@ void ompl_interface::ModelBasedPlanningContext::configure(const ros::NodeHandle&
+    ompl::base::ScopedState<> ompl_start_state(spec_.state_space_);
+    spec_.state_space_->copyToOMPLState(ompl_start_state.get(), getCompleteInitialRobotState());
+    ompl_simple_setup_->setStartState(ompl_start_state);
+    ompl_simple_setup_->setStateValidityChecker(ob::StateValidityCheckerPtr(new StateValidityChecker(this)));
```
Here too with the shared_ptr.  Please do a search through your pr for the keyword `new` and try to remove it in any place it could be a make_shared or make_unique.

JeroenDM replied:
I estimating there are close to 30 instances of this in the existing code base, but I think it will be confusing to change them all in this pull request. I changed these two instances to make_shared. But I'm not sure make_unique is ok here.

tylerjw commented on moveit_planners/ompl/ompl_interface/test/test_constrained_planning_state_space.cpp:
```diff
@@ -0,0 +1,330 @@
+{
+  testing::InitGoogleTest(&argc, argv);
+  return RUN_ALL_TESTS();
+}
```
newline at eof

JeroenDM replied:
done (a while ago)

tylerjw commented on moveit_planners/ompl/ompl_interface/test/test_constrained_state_validity_checker.cpp:
```diff
@@ -0,0 +1,326 @@
+#include <ompl/base/spaces/constraint/ProjectedStateSpace.h>
+
+/** \brief Use this flag to turn on extra output on std::cout for debugging. **/
+constexpr bool VERBOSE{ false };
```
here is another verbose flag

JeroenDM replied:
I will remove the `verbose` flags in all the other places. This one however is the same one as discussed in #2247 and is passed to the state validity checker. The comment will get updated once I rebase. (And I really should do this now, I'm going to have some nasty merge conflicts otherwise...)

JeroenDM replied:
Ooops, the comment will not update after the rebase unless I merge the different tests. I will update after the rebase.

JeroenDM replied:
done

tylerjw commented on moveit_planners/ompl/ompl_interface/test/test_constrained_state_validity_checker.cpp:
```diff
@@ -0,0 +1,326 @@
+    state_space_ = std::make_shared<ompl_interface::ConstrainedPlanningStateSpace>(space_spec);
+    state_space_->computeLocations();  // this gets called in the state space factory normally
+
+    assert(state_space_ != nullptr);
```
You don't need this.  If make_shared fails it throws a bad_alloc and the test will fail there. (that is assuming `computeLocations` won't reset state_space_ somehow?

JeroenDM replied:
done. `ComputeLocations` tries to access the memory but does not reset it as far as I understand.

tylerjw commented on moveit_planners/ompl/ompl_interface/test/test_constrained_state_validity_checker.cpp:
```diff
@@ -0,0 +1,326 @@
+    auto dummy_constraint = std::make_shared<DummyConstraint>(num_dofs_);
+    constrained_state_space_ = std::make_shared<ompl::base::ProjectedStateSpace>(state_space_, dummy_constraint);
+
+    assert(constrained_state_space_ != nullptr);
```
This too is extra, make_shared will throw if the allocation fails.

JeroenDM replied:
done

tylerjw commented on moveit_planners/ompl/ompl_interface/test/test_constrained_state_validity_checker.cpp:
```diff
@@ -0,0 +1,326 @@
+    // check if the we succeeded in the comment above
+    auto si = planning_context_spec_.ompl_simple_setup_->getSpaceInformation();
+    auto si_constrained = dynamic_cast<ompl::base::ConstrainedSpaceInformation*>(si.get());
+    assert(si_constrained != nullptr);
```
Use one of the gtest macros for this.

```suggestion
    ASSERT_NE(si_constrained, nullptr);
```

JeroenDM replied:
done

tylerjw commented on moveit_planners/ompl/ompl_interface/test/test_constrained_state_validity_checker.cpp:
```diff
@@ -0,0 +1,326 @@
+{
+  testing::InitGoogleTest(&argc, argv);
+  return RUN_ALL_TESTS();
+}
```
newline at eof

tylerjw commented on moveit_planners/ompl/ompl_interface/test/test_planning_context_manager.cpp:
```diff
@@ -0,0 +1,343 @@
+/***************************************************************************
+ * Run all tests on the Panda robot
+ * ************************************************************************/
+// class PandaTestPlanningContext : public TestPlanningContext
+// {
+// protected:
+//   PandaTestPlanningContext() : TestPlanningContext("panda", "panda_arm")
+//   {
+//   }
+// };
+
+// TEST_F(PandaTestPlanningContext, testSimpleRequest)
+// {
+//   // use the panda "ready" state from the srdf config as start state
+//   // we know this state should be within limits and self-collision free
+//   testSimpleRequest({ 0, -0.785, 0, -2.356, 0, 1.571, 0.785 }, { 0, -0.785, 0, -2.356, 0, 1.571, 0.685 });
+// }
+
+// TEST_F(PandaTestPlanningContext, testPathConstraints)
+// {
+//   testPathConstraints({ 0, -0.785, 0, -2.356, 0, 1.571, 0.785 }, { 0, -0.785, 0, -2.356, 0, 1.571, 0.685 });
+// }
```
Why are these commented out?

JeroenDM replied:
No reason, just stupidity. Thanks for commenting!

tylerjw commented on moveit_planners/ompl/ompl_interface/test/test_planning_context_manager.cpp:
```diff
@@ -0,0 +1,343 @@
+  }
+};
+
+// TEST_F(FanucTestPlanningContext, testSimpleRequest)
+// {
+//   testSimpleRequest({ 0, 0, 0, 0, 0, 0 }, { 0, 0, 0, 0, 0, 0.1 });
+// }
+
+// TEST_F(FanucTestPlanningContext, testPathConstraints)
+// {
+//   testPathConstraints({ 0, 0, 0, 0, 0, 0 }, { 0, 0, 0, 0, 0, 0.1 });
+// }
```
Here too?

JeroenDM replied:
Same as above. I uncommented them again.

tylerjw commented on moveit_planners/ompl/ompl_interface/test/test_planning_context_manager.cpp:
```diff
@@ -0,0 +1,343 @@
+  ros::init(argc, argv, "planning_context_manager_test");
+  testing::InitGoogleTest(&argc, argv);
+  return RUN_ALL_TESTS();
+}
```
[nit] newline at eof

tylerjw commented on moveit_planners/ompl/ompl_interface/test/test_planning_context_manager.test:
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0"?>
+<launch>
+    <test pkg="moveit_planners_ompl" type="test_planning_context_manager" test-name="test_planning_context_manager" />
+</launch>
```
[nit] newline at eof

tylerjw reviewed (commented):
I left some comments through this.  I didn't test running it as I didn't have time and don't know where to begin to do that.  Most of my comments are just C++ stuff that should be easy enough to clean up.  Thank you again for doing this, I look forward to getting this merged and seeing it work.

JeroenDM commented on moveit_planners/ompl/ompl_interface/src/parameterization/joint_space/constrained_planning_state_space.cpp:
```diff
@@ -0,0 +1,93 @@
+
+double* ompl_interface::ConstrainedPlanningStateSpace::getValueAddressAtIndex(ompl::base::State* ompl_state,
+                                                                              const unsigned int index) const
+{
```
It should be impossible for the end-user to misuse this part, as they do not directly handle state spaces normally.

The method also gets explicitly tested in `test_constrained_planning_state_space.cpp`, but that does not exclude that there could still be some bugs related to all this casting that is unfortunately necessary.

JeroenDM commented on moveit_planners/ompl/ompl_interface/src/parameterization/joint_space/constrained_planning_state_space_factory.cpp:
```diff
@@ -0,0 +1,58 @@
+  type_ = ConstrainedPlanningStateSpace::PARAMETERIZATION_TYPE;
+}
+
+int ompl_interface::ConstrainedPlanningStateSpaceFactory::canRepresentProblem(
+    const std::string& /*group*/, const moveit_msgs::MotionPlanRequest& /*req*/,
+    const moveit::core::RobotModelConstPtr& /*robot_model*/) const
+{
+  return -2;
+}
```
This is related to the rather complicated approach to select state spaces. It is documented in the corresponding header file:
```C++
  /** \brief Return a priority that this planner should be used for this specific planning problem.
   *
   * This state space factory is currently only used if `use_ompl_constrained_state_space` was set to `true` in
   * ompl_planning.yaml. In that case it is the only factory to choose from, so the priority does not matter.
   * It returns a low priority so it will never be choosen when others are available.
   * (The second lowest priority is -1 in the PoseModelStateSpaceFactory.)
   *
   * For more details on this state space selection process, see:
   * https://github.com/JeroenDM/moveit/pull/2
   * **/
```

I will add a comment in the source file that quickly explains it and point to the header file for more info.

JeroenDM replied:
Something like this:
```C++
  // Return the lowest priority currently existing in the ompl interface.
  // This state space will only be selected if it is the only option to choose from.
  // See header file for more info.
return -2;
```

tylerjw replied:
Ah, I'm sorry I didn't understand that.  Thank you for clarifying.

JeroenDM commented on moveit_planners/ompl/ompl_interface/src/parameterization/joint_space/constrained_planning_state_space.cpp:
```diff
@@ -0,0 +1,93 @@
+  return ompl_state->as<ompl_interface::ConstrainedPlanningStateSpace::StateType>()->values + index;
+}
+
+void ompl_interface::ConstrainedPlanningStateSpace::copyToRobotState(moveit::core::RobotState& robot_state,
+                                                                     const ompl::base::State* ompl_state) const
+{
+  robot_state.setJointGroupPositions(
+      spec_.joint_model_group_,
+      ompl_state->as<ompl::base::ConstrainedStateSpace::StateType>()->getState()->as<StateType>()->values);
+  robot_state.update();
+}
```
## Assert statement
I like the idea of adding `assert(ptr != nullptr);`. There are many other places in the existing OMPL interface where this could be valuable, but I feel like that does not belong in this PR. I'll just add it to the new code.

## `state->as<StateType>()` pattern
This pattern is used extremely often in the OMPL interface. It is based on the central design decision in OMPL to pass states around as pointers. Every function should cast the pointer to the proper type before using it. Because it is used so often, OMPL defines a helper method [State::<T>as()](https://ompl.kavrakilab.org/classompl_1_1base_1_1State.html#ace9774faf8793299dc105292f8557987) to make this easier.

For a constrained state space, it gets more complicated. We do not pass around `State` pointers, but a `WrapperState` (in full [ompl::base::WrapperStateSpace::StateType](https://ompl.kavrakilab.org/classompl_1_1base_1_1WrapperStateSpace_1_1StateType.html)). This type of state has a method `getState()` to access the "normal" state pointer. This "normal" state pointer is then cast again to the appropriate type. We end up with:

```C++
ompl_state->as<ompl::base::ConstrainedStateSpace::StateType>()->getState()->as<StateType>()
```
(This has caused me a lot of headaches (I mean, interesting challenges...), and I'm still not sure I completely understand all the implications.)

## Can we make this easier?
I think OMPL and MoveIt's OMPL interface are so tightly coupled that it is difficult to avoid the `state->as<StateType>()` pattern. But there could be ways to make it easier to use. Adding the assert statement seems like a good step in that direction.

## Q&A
> First is this an external function, do you expect users to use this or is it just for some other code you wrote?

This is all internal as far as I know. The user interface (ignoring MoveGroup for a moment) looks like this:

```C++
// setup the planning context manager
ompl_interface::PlanningContextManager pcm(robot_model_, constraint_sampler_manager_);

// get a planning context for a specific planning scene and request
auto pc = pcm.getPlanningContext(planning_scene_, request, error_code, node_handle_, false);

// solve your problem
planning_interface::MotionPlanResponse res;
bool res = pc->solve(res);
```
(The tests in `test_planning_context_manager.cpp` show this in more detail.)

> Second, if ompl_state was a const& can you use the as<>() functionality?

I'm not sure, can you static cast references? The implementation goes:

```C++
template <class T>
T *as()
 {
 BOOST_CONCEPT_ASSERT((boost::Convertible<T *, State *>));
   return static_cast<T *>(this);
}
```

Also, many of the functions that take raw state pointers implement an interface imposed by OMPL. (Not the one discussed here though.)

tylerjw replied:
It looks like there is a boost assert there so we shouldn't need to add ours.  (I'm not sure what a cncept assert is).

Looking at the examples (the tests) I don't mind this interface nearly as much.  It seems like it is this way because of the interface of ompl and the intersection with the moveit stadards (the ref to the RobotState).  I think the combination of a non-const ref and a const ptr had me confused what you were trying to do here.

tylerjw replied:
I do think adding `assert(ptr != nullptr);` would be useful.  The assert I don't think we need is one that tests if the cast will work (as that is already in the as<>() function).

JeroenDM replied:
The assert statements are added.

JeroenDM commented on moveit_planners/ompl/ompl_interface/src/parameterization/joint_space/constrained_planning_state_space.cpp:
```diff
@@ -0,0 +1,93 @@
+  robot_state.update();
+}
+
+void ompl_interface::ConstrainedPlanningStateSpace::copyToOMPLState(ompl::base::State* ompl_state,
+                                                                    const moveit::core::RobotState& robot_state) const
```
:+1:

JeroenDM commented on moveit_planners/ompl/ompl_interface/src/model_based_planning_context.cpp:
```diff
@@ -116,11 +117,23 @@ void ompl_interface::ModelBasedPlanningContext::configure(const ros::NodeHandle&
+    spec_.state_space_->copyToOMPLState(ompl_start_state.get(), getCompleteInitialRobotState());
+    ompl_simple_setup_->setStartState(ompl_start_state);
+    ompl_simple_setup_->setStateValidityChecker(
+        ob::StateValidityCheckerPtr(new ConstrainedPlanningStateValidityChecker(this)));
```
Note to self @JeroenDM , did `std::make_shared` work here? I seem to remember it did not work in some places.

JeroenDM replied:
I changed it and the tests pass.

JeroenDM commented on moveit_planners/ompl/ompl_interface/src/planning_context_manager.cpp:
```diff
@@ -90,7 +96,7 @@ ompl_interface::MultiQueryPlannerAllocator::~MultiQueryPlannerAllocator()
   // Store all planner data
   for (const auto& entry : planner_data_storage_paths_)
   {
-    ROS_INFO("Storing planner data");
+    ROS_INFO_NAMED(LOGNAME, "Storing planner data");
```
This is not my code, but I feel an info message is justified. You have to ask MoveIt explicitly to reuse a planning graph, so it seems nice to have some feedback that it actually happened, no? See #1799 for the related PR.

tylerjw replied:
:+1:

JeroenDM commented:
I just did the scariest rebase so far :) Now I can continue with addressing the review comments. The biggest chunk left is refactoring the bounds class. Also, there are some smaller things that I can clean up in the tests based on the feedback on my other pull requests that got merged recently.

mamoll commented on moveit_planners/ompl/ompl_interface/include/moveit/ompl_interface/detail/ompl_constraints.h:
```diff
@@ -93,6 +93,47 @@ struct Bounds
+ * Or you can use the EqualityPositionConstraint version by setting the name of the constraint to
+ * "use_equality_constraints". But the latter will ignore bounds on other dimensions.
+ * **/
+class Bounds2
```
```suggestion
class Bounds
```

mamoll replied:
Maybe you are still working on this, but I assume that eventually you will rename Bounds2 to Bounds and eliminate the Bounds class above, right?

JeroenDM replied:
Indeed, I am committing work in progress. Sorry for the confusion, I'll let you know when the refactoring is finished.

mamoll commented on moveit_planners/ompl/ompl_interface/CMakeLists.txt:
```diff
@@ -49,14 +52,25 @@ if(CATKIN_ENABLE_TESTING)
-  target_link_libraries(test_planning_context_manager ${MOVEIT_LIB_NAME} ${OMPL_LIBRARIES} ${catkin_LIBRARIES} ${eigen_conversions_LIBRARIES})
-
+  target_link_libraries(test_planning_context_manager ${MOVEIT_LIB_NAME} ${OMPL_LIBRARIES} ${catkin_LIBRARIES})
+  
```
```suggestion

```

JeroenDM replied:
This change must have sneaked in when solving the merge conflicts. I linked the library again, it is required for the test. Or did you mean something else with the suggestions?

mamoll replied:
I just suggested deleting the trailing whitespace. Obviously, if `${eigen_conversions_LIBRARIES}` is still needed, then don't delete that :-).

JeroenDM replied:
Ah, ok. I have a single whitespace before every "add test" command now.

mamoll commented on moveit_planners/ompl/ompl_interface/include/moveit/ompl_interface/detail/ompl_constraints.h:
```diff
@@ -0,0 +1,325 @@
+{
+public:
+  BaseConstraint(const robot_model::RobotModelConstPtr& robot_model, const std::string& group,
+                 const unsigned int num_dofs, const unsigned int num_cons_ = 3);
```
Also, remove the trailing underscore since this is typically used for member variables.

JeroenDM replied:
I've added a comment. All constraints currently implemented have 3 constraint equations, so it might just as well be the default. The MoveIt constraint messages such as position and orientation constraints also have 3 dimensions.

JeroenDM commented on moveit_planners/ompl/ompl_interface/src/planning_context_manager.cpp:
```diff
@@ -343,14 +350,42 @@ ompl_interface::ModelBasedPlanningContextPtr ompl_interface::PlanningContextMana
+    }
 
     ROS_DEBUG_NAMED(LOGNAME, "Creating new planning context");
     context.reset(new ModelBasedPlanningContext(config.name, context_spec));
```
See earlier comment
>I estimate there are close to 30 instances of this in the existing code base, but I think it will be confusing to change them all in this pull request

@JeroenDM add this in a separate pull request?

JeroenDM commented:
**Update** I am slowly working my way through the review comments and the `Bounds` refactor is in progress. (I'm working on it a little bit in the morning before work.) In any case, thank you already for the many in-depth review comments and patience, I will get there! :)

JeroenDM commented on moveit_planners/ompl/ompl_interface/src/detail/ompl_constraints.cpp:
```diff
@@ -0,0 +1,343 @@
+void BaseConstraint::jacobian(const Eigen::Ref<const Eigen::VectorXd>& joint_values,
+                              Eigen::Ref<Eigen::MatrixXd> out) const
+{
+  Eigen::VectorXd constraint_error = calcError(joint_values);
```
A lot of temporary objects. Can I avoid them?

JeroenDM commented on moveit_planners/ompl/ompl_interface/src/parameterization/joint_space/constrained_planning_state_space.cpp:
```diff
@@ -0,0 +1,93 @@
+  ompl_state->as<ompl::base::ConstrainedStateSpace::StateType>()->getState()->as<StateType>()->clearKnownInformation();
+}
+
+void ompl_interface::ConstrainedPlanningStateSpace::copyJointToOMPLState(ompl::base::State* ompl_state,
+                                                                         const moveit::core::RobotState& robot_state,
+                                                                         const moveit::core::JointModel* joint_model,
+                                                                         int ompl_state_joint_index) const
+{
```
Resolved based on an earlier discussion:

>Looking at the examples (the tests) I don't mind this interface nearly as much. It seems like it is this way because of the interface of ompl and the intersection with the moveit stadards (the ref to the RobotState).

JeroenDM commented on moveit_planners/ompl/ompl_interface/src/planning_context_manager.cpp:
```diff
@@ -164,7 +170,7 @@ ompl::base::PlannerPtr ompl_interface::MultiQueryPlannerAllocator::allocatePlann
   // Try to initialize planner with loaded planner data
   if (load_planner_data)
   {
-    ROS_INFO("Loading planner data");
+    ROS_INFO_NAMED(LOGNAME, "Loading planner data");
```
Resolved because of my comment higher up in the code:

>This is not my code, but I feel an info message is justified. You have to ask MoveIt explicitly to reuse a planning graph, so it seems nice to have some feedback that it actually happened, no? See #1799 for the related PR.

JeroenDM commented on moveit_planners/ompl/ompl_interface/src/planning_context_manager.cpp:
```diff
@@ -296,6 +302,7 @@ void ompl_interface::PlanningContextManager::registerDefaultStateSpaces()
 {
   registerStateSpaceFactory(ModelBasedStateSpaceFactoryPtr(new JointModelStateSpaceFactory()));
   registerStateSpaceFactory(ModelBasedStateSpaceFactoryPtr(new PoseModelStateSpaceFactory()));
+  registerStateSpaceFactory(ModelBasedStateSpaceFactoryPtr(new ConstrainedPlanningStateSpaceFactory()));
```
As mentioned below, there are too many cases in the existing code to add this to the current PR I think. I added to my TODO list to create a separate PR.

JeroenDM commented on moveit_planners/ompl/ompl_interface/src/planning_context_manager.cpp:
```diff
@@ -343,14 +350,42 @@ ompl_interface::ModelBasedPlanningContextPtr ompl_interface::PlanningContextMana
-      std::unique_lock<std::mutex> slock(cached_contexts_->lock_);
-      cached_contexts_->contexts_[std::make_pair(config.name, factory->getType())].push_back(context);
+      {
+        std::unique_lock<std::mutex> slock(cached_contexts_->lock_);
```
I'm not entirely sure how it works with the cached context having a persistent lock, but as there isn't any `unlock()` call anywhere in the code, I agree with replacing it with the less complex default option.

JeroenDM commented:
@mamoll The bound refactor is done. There is still some less ideal code with many temporary objects:
```C++
void BaseConstraint::jacobian(const Eigen::Ref<const Eigen::VectorXd>& joint_values,
                              Eigen::Ref<Eigen::MatrixXd> out) const
{
  Eigen::VectorXd constraint_error = calcError(joint_values);
  Eigen::VectorXd constraint_derivative = bounds_.derivative(constraint_error);
  Eigen::MatrixXd robot_jacobian = calcErrorJacobian(joint_values);
  for (std::size_t i{ 0 }; i < bounds_.size(); ++i)
  {
    out.row(i) = constraint_derivative[i] * robot_jacobian.row(i);
  }
}
```
The loop could potentially be replaced with Eigen magic, but I'm not sure it's worth it performance or readability wise.

JeroenDM commented:
There are some warnings and clang-format things I need to take care of on Travis. The build for the tests times out.

mamoll commented:
@JeroenDM There are still errors in the CI, some clang-format related, some related to other tests. I don't see the time outs. Can you have another look?

JeroenDM commented:
@mamoll I was confused because this branch was built on my personal Travis, I'm not sure why this happened.

Sorry about the clang-format issues again, they should be fixed now.

The following tests are failing:
- (noetic / melodic) moveit_core/gtest-test_bullet_continuous_collision_checking.xml
- (noetic / melodic) moveit_ros_planning_interface/rosunit-subframes_test.xml
- (melodic) moveit_ros_planning_interface/nosetests-serialize_msg.py.xml

They seem unrelated to this PR.

JeroenDM commented:
There is still a hack left in the code for equality constraints, but I don't have a better solution for now. Changing the constraint message definition could be a solution. Maybe it's ok to leave this for a future fix?

@mamoll @felixvd 

I added a (long...) explanation in the code and changed it to a static const member as Mark pointed out in [this comment](https://github.com/ros-planning/moveit/pull/2273#discussion_r504790877).

```C++
  /** \brief Position bounds under this threshold are interpreted as equality constraints, the others as unbounded.
   *
   * WARNING: Below follows the explanation of an ugly hack. Ideally, the user could specify equality constraints by
   * setting the constraint dimension to zero. However, this would result in a constraint region primitive with a zero
   * dimension in MoveIt, which the planner can (almost) never satisfy. Therefore we use a threshold value, below which
   * the constraint is interpreted as an equality constraint. This threshold value is not used in the planning algorithm itself!
   *
   * This threshold value should be larger than the tolerance of the constraints specified in OMPL
   * (ompl::magic::CONSTRAINT_PROJECTION_TOLERANCE = 1e-4).
   *
   * This is necessary because the constraints are also checked by MoveIt in the StateValidity checker. If this check
   * would use a stricter tolerance than was used to satisfy the constraints in OMPL, all states would be invalid.
   * Therefore the dimension of an equality constraint specified in the constraint message should be at least equal the
   * the tolerance used by OMPL to project onto the constraints. To avoid checking for exact equality on floats, the
   * threshold is made a bit larger.
   *
   * equality_constraint_threshold_ > tolerance in constraint message > OMPL projection tolerance
   *
   * That's why the value is 1e-3 > 1e-4.
   * Now the user can specify any value between 1e-3 and 1e-4 to specify an equality constraint.
   * **/
  static constexpr double EQUALITY_CONSTRAINT_THRESHOLD_{ 0.001 };
```

JeroenDM commented on moveit_planners/ompl/ompl_interface/include/moveit/ompl_interface/detail/ompl_constraints.h:
```diff
@@ -0,0 +1,325 @@
+private:
+  /** \brief Position bounds under this threshold are interpreted as equality constraints, the others as unbounded.
+   *
+   * This threshold value should be larger that the tolerance of the constraints specificied in ompl
```
I changed it to static const and continued the discussion on why it exists and how large it should be in the main thread to make it easier to find.

JeroenDM commented on moveit_planners/ompl/ompl_interface/test/test_planning_context_manager.cpp:
```diff
@@ -0,0 +1,343 @@
+#include <moveit/ompl_interface/parameterization/joint_space/constrained_planning_state_space.h>
+
+/** \brief Use this flag to turn on extra output on std::cout for debugging. **/
+// constexpr bool VERBOSE{ true };
```
I removed the verbose flags where possible. (Some are necessary because the state validity checker uses an explicit verbose flag.)

JeroenDM commented on moveit_planners/ompl/ompl_interface/test/test_constrained_state_validity_checker.cpp:
```diff
@@ -0,0 +1,326 @@
+    planning_context_->setCompleteInitialState(*initial_robot_state_);
+
+    if (VERBOSE)
+      ROS_INFO("Planning context with name %s is ready (but not configured).", planning_context_->getName().c_str());
```
done

JeroenDM commented on moveit_planners/ompl/ompl_interface/include/moveit/ompl_interface/detail/ompl_constraints.h:
```diff
@@ -0,0 +1,325 @@
+   * */
+  virtual Eigen::VectorXd calcError(const Eigen::Ref<const Eigen::VectorXd>& x) const
+  {
+    ROS_ERROR_STREAM("Constraint method calcError was not overridded, so it should not be used.");
```
done

JeroenDM commented on moveit_planners/ompl/ompl_interface/src/detail/state_validity_checker.cpp:
```diff
@@ -194,4 +196,112 @@ double ompl_interface::StateValidityChecker::clearance(const ompl::base::State*
+  const kinematic_constraints::KinematicConstraintSetPtr& kset = planning_context_->getPathConstraints();
+  if (kset && !kset->decide(*robot_state, verbose).satisfied)
+  {
+    const_cast<ob::State*>(state)->as<ModelBasedStateSpace::StateType>()->markInvalid();
```
I understand that this an (arguably less ideal) way to implement caching of the validity. I added a single comment at the top of the file to avoid copying the same comment many times:
```C++
/** A state validity checker checks:
 *
 * - Bounds (joint limits).
 * - Collision.
 * - Kinematic path constraints.
 * - Generic user-specified feasibility using the `isStateFeasible` of the planning scene.
 *
 * IMPORTANT: Although the isValid method takes the state as `const ompl::base::State* state`,
 * it uses const_cast to modify the validity of the state with `markInvalid` and `markValid` for caching.
 * **/
```

gleichdick replied:
We could also declare the `flags` member of `ModelBasedStateSpace::StateType` as `mutable`, so that it can be changed from a `const` qualified member function. I'm happy to provide a PR.
But one question (as I don't know the details): is the state a shared object between threads? If so, the `flags` should probably be an `atomic` to avoid data races.

JeroenDM commented:
As you can see in the latest commits I processed some of the minor review comments that were leftover.

I think the essential stuff is OK now and it is ready to be merged, apart from the following CI failures related bullet (just skipped tests) subframes (TestPlanUsingSubframes) and move_group_interface. The latter is the most suspicious in this context:

```
<testsuites tests="5" failures="1" disabled="0" errors="0" timestamp="2020-12-11T13:44:22" time="27.937" name="AllTests">
  <testsuite name="MoveGroupTestFixture" tests="5" failures="1" disabled="0" errors="0" time="27.937">
    <testcase name="MoveToPoseTest" status="run" time="4.534" classname="MoveGroupTestFixture" />
    <testcase name="JointSpaceGoalTest" status="run" time="3.8" classname="MoveGroupTestFixture" />
    <testcase name="PathConstraintTest" status="run" time="6.301" classname="MoveGroupTestFixture" />
    <testcase name="CartPathTest" status="run" time="7.709" classname="MoveGroupTestFixture" />
    <testcase name="CollisionObjectsTest" status="run" time="5.593" classname="MoveGroupTestFixture">
      <failure message="/root/ros_ws/src/moveit/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp:100&#x0A;      Expected: move_group_-&gt;plan(my_plan)&#x0A;      Which is: 4-byte object &lt;FE-FF FF-FF&gt;&#x0A;To be equal to: moveit::planning_interface::MoveItErrorCode::SUCCESS&#x0A;      Which is: 1&#x0A;Google Test trace:&#x0A;/root/ros_ws/src/moveit/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp:98: planAndMove&#x0A;/root/ros_ws/src/moveit/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp:91: planAndMoveToPose&#x0A;/root/ros_ws/src/moveit/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp:286: CollisionObjectsTest" type=""><![CDATA[/root/ros_ws/src/moveit/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp:100
      Expected: move_group_->plan(my_plan)
      Which is: 4-byte object <FE-FF FF-FF>
To be equal to: moveit::planning_interface::MoveItErrorCode::SUCCESS
      Which is: 1
Google Test trace:
/root/ros_ws/src/moveit/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp:98: planAndMove
/root/ros_ws/src/moveit/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp:91: planAndMoveToPose
/root/ros_ws/src/moveit/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp:286: CollisionObjectsTest]]></failure>
      <failure message="/root/ros_ws/src/moveit/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp:109&#x0A;Value of: actual.isApprox(expected, EPSILON)&#x0A;  Actual: false&#x0A;Expected: true&#x0A;expected: &#x0A;   1    0    0  0.4&#x0A;   0    1    0 -0.4&#x0A;   0    0    1  0.7&#x0A;   0    0    0    1&#x0A;actual: &#x0A;           1  1.75267e-08 -8.58801e-07         0.28&#x0A; -1.7526e-08            1  7.60527e-07         -0.2&#x0A; 8.58801e-07 -7.60527e-07            1          0.5&#x0A;           0            0            0            1&#x0A;Google Test trace:&#x0A;/root/ros_ws/src/moveit/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp:106: testEigenPose&#x0A;/root/ros_ws/src/moveit/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp:114: testPose(const Eigen::Isometry3d&amp;)&#x0A;/root/ros_ws/src/moveit/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp:126: testPose(const geometry_msgs::Pose&amp;)&#x0A;/root/ros_ws/src/moveit/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp:286: CollisionObjectsTest" type=""><![CDATA[/root/ros_ws/src/moveit/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp:109
Value of: actual.isApprox(expected, EPSILON)
  Actual: false
Expected: true
expected: 
   1    0    0  0.4
   0    1    0 -0.4
   0    0    1  0.7
   0    0    0    1
actual: 
           1  1.75267e-08 -8.58801e-07         0.28
 -1.7526e-08            1  7.60527e-07         -0.2
 8.58801e-07 -7.60527e-07            1          0.5
           0            0            0            1
```

gleichdick reviewed (commented):
There is an option to get rid of the `const_cast` of the `ompl::Base::state` subclass:

bostoncleek commented:
> As you can see in the latest commits I processed some of the minor review comments that were leftover.
> 
> I think the essential stuff is OK now and it is ready to be merged, apart from the following CI failures related bullet (just skipped tests) subframes (TestPlanUsingSubframes) and move_group_interface. The latter is the most suspicious in this context:
> 
> ```
> <testsuites tests="5" failures="1" disabled="0" errors="0" timestamp="2020-12-11T13:44:22" time="27.937" name="AllTests">
>   <testsuite name="MoveGroupTestFixture" tests="5" failures="1" disabled="0" errors="0" time="27.937">
>     <testcase name="MoveToPoseTest" status="run" time="4.534" classname="MoveGroupTestFixture" />
>     <testcase name="JointSpaceGoalTest" status="run" time="3.8" classname="MoveGroupTestFixture" />
>     <testcase name="PathConstraintTest" status="run" time="6.301" classname="MoveGroupTestFixture" />
>     <testcase name="CartPathTest" status="run" time="7.709" classname="MoveGroupTestFixture" />
>     <testcase name="CollisionObjectsTest" status="run" time="5.593" classname="MoveGroupTestFixture">
>       <failure message="/root/ros_ws/src/moveit/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp:100&#x0A;      Expected: move_group_-&gt;plan(my_plan)&#x0A;      Which is: 4-byte object &lt;FE-FF FF-FF&gt;&#x0A;To be equal to: moveit::planning_interface::MoveItErrorCode::SUCCESS&#x0A;      Which is: 1&#x0A;Google Test trace:&#x0A;/root/ros_ws/src/moveit/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp:98: planAndMove&#x0A;/root/ros_ws/src/moveit/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp:91: planAndMoveToPose&#x0A;/root/ros_ws/src/moveit/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp:286: CollisionObjectsTest" type=""><![CDATA[/root/ros_ws/src/moveit/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp:100
>       Expected: move_group_->plan(my_plan)
>       Which is: 4-byte object <FE-FF FF-FF>
> To be equal to: moveit::planning_interface::MoveItErrorCode::SUCCESS
>       Which is: 1
> Google Test trace:
> /root/ros_ws/src/moveit/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp:98: planAndMove
> /root/ros_ws/src/moveit/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp:91: planAndMoveToPose
> /root/ros_ws/src/moveit/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp:286: CollisionObjectsTest]]></failure>
>       <failure message="/root/ros_ws/src/moveit/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp:109&#x0A;Value of: actual.isApprox(expected, EPSILON)&#x0A;  Actual: false&#x0A;Expected: true&#x0A;expected: &#x0A;   1    0    0  0.4&#x0A;   0    1    0 -0.4&#x0A;   0    0    1  0.7&#x0A;   0    0    0    1&#x0A;actual: &#x0A;           1  1.75267e-08 -8.58801e-07         0.28&#x0A; -1.7526e-08            1  7.60527e-07         -0.2&#x0A; 8.58801e-07 -7.60527e-07            1          0.5&#x0A;           0            0            0            1&#x0A;Google Test trace:&#x0A;/root/ros_ws/src/moveit/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp:106: testEigenPose&#x0A;/root/ros_ws/src/moveit/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp:114: testPose(const Eigen::Isometry3d&amp;)&#x0A;/root/ros_ws/src/moveit/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp:126: testPose(const geometry_msgs::Pose&amp;)&#x0A;/root/ros_ws/src/moveit/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp:286: CollisionObjectsTest" type=""><![CDATA[/root/ros_ws/src/moveit/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp:109
> Value of: actual.isApprox(expected, EPSILON)
>   Actual: false
> Expected: true
> expected: 
>    1    0    0  0.4
>    0    1    0 -0.4
>    0    0    1  0.7
>    0    0    0    1
> actual: 
>            1  1.75267e-08 -8.58801e-07         0.28
>  -1.7526e-08            1  7.60527e-07         -0.2
>  8.58801e-07 -7.60527e-07            1          0.5
>            0            0            0            1
> ```

After running this test multiple times, sometimes it fails and other times it does not. The failure could be result of the way the test is written. It may not be because of this feature.

JeroenDM commented:
@gleichdick 
> We could also declare the flags member of ModelBasedStateSpace::StateType as mutable, so that it can be changed from a const qualified member function. I'm happy to provide a PR.
But one question (as I don't know the details): is the state a shared object between threads? If so, the flags should probably be an atomic to avoid data races.

The state validity checker should be thread save, so we should make the flags `atomic` indeed.

A pull request would be great! Then people with more knowledge of these C++ tricks than me can review it :) There is a big chance it will get merged faster than this pull request and I can just rebase to use it. If it's the other way around, then you can rebase :)

JeroenDM commented:
@bostoncleek 
> After running this test multiple times, sometimes it fails and other times it does not. The failure could be result of the way the test is written. It may not be because of this feature.

Is it also occurring in other PR's? I've not looked into this yet.

@mamoll  Also, I've created backports to melodic because I was personally interested in testing it on Ubuntu 18.
- [this branch for melodic](https://github.com/JeroenDM/moveit/tree/ompl-constrained-planning-melodic)
- [tutorial branch for melodic](https://github.com/JeroenDM/moveit_tutorials/tree/ompl-constrained-planning-melodic)

bostoncleek commented:
I merged a test [PR #2484](https://github.com/ros-planning/moveit/pull/2484) that should resolve the test failure above.

@JeroenDM  Can you rebase your PR against master?

JeroenDM commented:
@bostoncleek 
I just did the rebase but I had to go through a minefield of merge conflicts, so I hope I did not break anything.

Tomorrow I will manually read through the tests again just to double-check that all the new tests are actually executed because those, in particular, were sketchy to combine :)

@tylerjw I'm probably ... really should clean up the commit history, as squash committing 2500 lines sounds terrible...
(To create the melodic backport, I added all the code manually (careful copy-pasting :) ) and this was actually reasonably fast to do, but that would maybe require a review again...)

tylerjw commented:
Please rebase again.  @bostoncleek fixed the other test that is blocking this and we just merged it.

tylerjw commented:
So for the subframes test, I did some testing locally, and here is some of the debug log when it fails (it is not consistently passing on master):

```
ros.moveit_planners_ompl.ompl: panda_arm/panda_arm: Skipping invalid start state (invalid state)
ros.moveit_kinematics.kdl: IK timed out after 0.0761564 > 0.05s and 1 attempts
ros.moveit_planners_ompl.ompl: panda_arm/panda_arm: Discarded start state panda_joint1 = 0 
panda_joint2 = -0.785 
panda_joint3 = 0 
panda_joint4 = -2.356 
panda_joint5 = 0 
panda_joint6 = 1.571 
panda_joint7 = 0.785 
* invalid state
Tag: -1

ros.moveit_planners_ompl.ompl: panda_arm/panda_arm: Motion planning start tree could not be initialized!
ros.moveit_planners_ompl.ompl: No solution found after 0.000865 seconds
ros.moveit_planners_ompl.ompl: Attempting to stop goal sampling thread...
ros.moveit_planners_ompl.ompl: Goal sampling thread never did any work.
ros.moveit_planners_ompl.ompl: Stopped goal sampling thread after 0 sampling attempts
ros.moveit_planners_ompl.model_based_planning_context: There were 0 valid motions and 0 invalid motions.
ros.moveit_planners_ompl.model_based_planning_context: Unable to solve the planning problem
ros.moveit_ros_planning_interface.move_group_interface: Fail: ABORTED: No motion plan found. No execution attempted.
ros.moveit_ros_planning_interface: Failed to perform motion.
/home/tyler/code/moveit/ws_moveit/src/moveit/moveit_ros/planning_interface/test/subframes_test.cpp:170: Failure
Value of: moveToCartPose(target_pose_stamped, group, "cylinder/tip")
  Actual: false
Expected: true
[  FAILED  ] TestPlanUsingSubframes.SubframesTests (2483 ms)
[----------] 1 test from TestPlanUsingSubframes (2483 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (2483 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] TestPlanUsingSubframes.SubframesTests

 1 FAILED TEST```

JeroenDM commented:
Another rebase will be necessary after #2472 I suspect.

davetcoleman commented:
Do I understand correctly that this PR got merged into MoveIt 2, but is still open here for MoveIt 1? Do we want to abandon this effort for MoveIt 1? @mamoll @JeroenDM

JeroenDM commented:
(Also, I will rebase it again and maybe the CI finally gives us a green light.)

davetcoleman commented:
I've opened this PR against MoveIt2:
https://github.com/ros-planning/moveit2_tutorials/pull/87

JafarAbdi commented on moveit_planners/ompl/ompl_interface/src/detail/threadsafe_state_storage.cpp:
```diff
@@ -61,6 +61,7 @@ moveit::core::RobotState* ompl_interface::TSStateStorage::getStateStorage() cons
   if (it == thread_states_.end())
   {
     st = new moveit::core::RobotState(start_state_);
+    st->setToDefaultValues();  // avoid uninitialized memory
```
FYI, this line introduced a regression on moveit2 https://github.com/ros-planning/moveit2/pull/455

v4hn commented:
Let's close this as #2742 continues these efforts and adds on them.

JeroenDM commented:
@v4hn could you maybe reopen this PR for CI? I'm going to try the same rebase as in  #2742, but with more cherry picking and manual coding to avoid the issues that occurred there. I think this might be faster than fixing  #2742... not sure.

v4hn commented:
You should have been able to do that yourself? Anyway I opened it, but it's even conflicting at the moment.

JeroenDM commented:
>You should have been able to do that yourself? 

Sorry, I was logged out for some reason and did not see it afterwards.

The subframes test is failing, I'm not sure if this can be related to this PR... @felixvd 
```
 Full test results for 'build/moveit_ros_planning_interface/test_results/moveit_ros_planning_interface/rosunit-subframes_test.xml'
-------------------------------------------------
<?xml version="1.0" encoding="UTF-8"?>
<testsuites tests="1" failures="1" disabled="0" errors="0" timestamp="2021-07-07T13:17:15" time="1.274" name="AllTests">
  <testsuite name="TestPlanUsingSubframes" tests="1" failures="1" disabled="0" errors="0" time="1.274">
    <testcase name="SubframesTests" status="run" time="1.274" classname="TestPlanUsingSubframes">
      <failure message="/home/runner/work/moveit/moveit/.work/target_ws/src/moveit/moveit_ros/planning_interface/test/subframes_test.cpp:173&#x0A;Value of: moveToCartPose(target_pose_stamped, group, &quot;cylinder/tip&quot;)&#x0A;  Actual: false&#x0A;Expected: true" type=""><![CDATA[/home/runner/work/moveit/moveit/.work/target_ws/src/moveit/moveit_ros/planning_interface/test/subframes_test.cpp:173
Value of: moveToCartPose(target_pose_stamped, group, "cylinder/tip")
  Actual: false
Expected: true]]></failure>
    </testcase>
  </testsuite>
</testsuites>
```

I was reading through the commits that I would cherry pick to add orientation constraints support, and I really think we should do this is a separate PR. The code is this PR has been reviewed, but the new code would not be. Creating a separate PR with a limited amount of code is much easier to review.

gautz commented:
@JeroenDM not sure it is related to the subframes test, but I cherry-picked from my PR the commit requested by @v4hn . See discussion https://github.com/ros-planning/moveit/pull/2742#issuecomment-872477267

> I was reading through the commits that I would cherry pick to add orientation constraints support,
Can't wait to see this too :D

JeroenDM commented:
@mamoll I think all comments from your old review got resolved.

It could be useful to have a last review from someone else, although I think all lines of code have been reviewed at some point. Given that the test coverage is not terrible it can also be good to go as is.

gautz commented:
I don't understand why @mamoll requested changes do not appear as resolved.
Maybe @zkingston could have a look here too?

gautz commented:
@JeroenDM I see planning_context_manager.cpp is conflicting. Could you have a look?

JeroenDM commented:
@gautz I think the conflict should be fixed now. It was related to the `new`'s that were replaced by `std::make_shared` recently. I used the online interface, hope it is formatted correctly and works.

gautz commented on moveit_planners/ompl/ompl_interface/src/planning_context_manager.cpp:
```diff
@@ -372,11 +372,11 @@ ompl_interface::ModelBasedPlanningContextPtr ompl_interface::PlanningContextMana
 
     ROS_DEBUG_NAMED(LOGNAME, "Creating new planning context");
-    context.reset(new ModelBasedPlanningContext(config.name, context_spec));
+    context_spec.ompl_simple_setup_ = std::make_shared<ompl::geometric::SimpleSetup>(context_spec.state_space_);
```
the modernization with std::make_shared introduced a regression. With these two line changes I cannot run the Constrained planning tutorial anymore, it is crashing.

JeroenDM replied:
:( I remember similar problems last year. Do you have some error messages?

gautz replied:
unfortunately it is just crashing.

JeroenDM replied:
Jup this was clearly a bug, my mistake :s I ran the tutorial locally again now and it seems to be working again. (Without the support for orientation constraints with this planner of course, but it just falls back to the PoseModel planner.) (FWI turning on debug messages can be really helpful http://dav.ee/blog/notes/archives/898 :) )

gautz commented:
I looked a bit at the moveit2 PR, the core functionalities (/ompl_interface*) are just copied with some small modification. I feel like most of the changes are just related to the port to ROS2, changing namespaces. Also clang was run which makes the PR much bigger than this one. The only visible review/modifications/bugfixes are also mostly ROS2 specific, e.g. https://github.com/ros-planning/moveit2/pull/347#pullrequestreview-583593269.  
So porting back a port does sound like a lots more overhead than maybe erasing small diffs from each core files.

I did not check what was done with the tests and the demos though. What is your opinion @JeroenDM , especially on the tests?

Would be nice to have thoughts of @bostoncleek on this also. Were there any bug fixes independent from the ROS2 migration?

JeroenDM commented:
I'm sorry I have so little time and did not really add anything useful in the maintainer meeting.

But if @gautz analysis is correct
> I looked a bit at the moveit2 PR, the core functionalities (/ompl_interface*) are just copied with some small modifications. I feel like most of the changes are just related to the port to ROS2, changing namespaces.

and given that backporting from moveit2 to moveit1 does not seem hassle-free anyway, I cannot see major disadvantages in merging this PR instead of *someone* trying to backport the moveit2 one.

That is, given that I fix this recent issue that should have been an innocent merge conflict fix. I will look into this tomorrow.
It turns out it is difficult to revert the specific changes that I did when resolving the merge conflict, except manually changing the lines.

gautz commented:
@JeroenDM could you check the tests and the demos from the MoveIt2 PR to see if there are important things we should backport to this PR?

gautz commented:
https://github.com/ros-planning/moveit/issues/2861

fmessmer commented:
@JeroenDM @gautz 
I wonder whether there is still effort going on in getting this merged into moveit(1) or whether focus is on extending this feature in moveit2 (ref https://github.com/ros-planning/moveit2/pull/347)
it's a bit tricky from the outside to assess the current state and what's missing for the merge

gautz commented:
The last idea from MoveIt maintainers was to backport the MoveIt2 PR. Which I did not agree (see previous message):
> porting back a port does sound like a lots more overhead 

So remaining ToDos may be:
- erasing small diffs (between this PR and [moveit2#347](https://github.com/ros-planning/moveit2/pull/347)) from each core files.
- Backport the tests, for this I asked @JeroenDM to have a look

But my general feeling was that no MoveIt1 maintainer has time to work on this.
Thus if I find time in the last 3 months of my current position, I'll try 
> extending this feature in moveit2 (ref [ros-planning/moveit2#347](https://github.com/ros-planning/moveit2/pull/347))

v4hn commented:
I believe there is nothing to see "from the inside" of the maintainer group that you can't see in the issues.

It would be great to see this feature working in MoveIt out of the box and I agree that backporting the MoveIt2 PR is not a good idea as multiple people stated it was merged prematurely.
I don't have a concrete use-case for this myself at the moment and it is too involved to look into it on the side. This seems to be the case for other maintainers as well.

Any work towards merging and/or improving these features is very welcome!

captain-yoshi commented:
@v4hn @gautz  Would this project be a good candidate for GSoC22 ? Maybe as a Medium project. Don't really know how much work is left...

v4hn commented:
this was a gsoc project already. thus I do not consider this as a viable candidate unless the scope is significant extended.

knorth55 commented:
we have this issue in our Fetch robot.
what is the status of this issue?
